File: BOJ/17127.cpp
/*

벚나무의 개수 N이 최대 10이므로 brutefoce를 사용했다.
4개의 그룹으로 나누는 모든 경우의 수마다 값을 구하고 가장 큰 값을 answer에 저장해주었다.

*/
#include<bits/stdc++.h>
using namespace std;

int answer = 0;
int N;
int sakura[10];
int check[10];

// init: 그룹의 시작 지점, cnt: 그룹을 나눈 횟수, sum: 네 개의 P의 최대값을 구하기 위해 그룹을 나눌 때마다 더해준다.
void bruteForce(int init, int cnt, int sum) {
	int tmp = 1;

	// 그룹을 3번 나누게 되면 4개의 그룹이 되므로 기저조건으로 잡았다.
	if (cnt == 3) {
		bool flag = false;
		for (int i = init; i < N; i++) {
			flag = true; // 4개의 그룹을 정확히 나누었는지 체크해준다.
			tmp *= sakura[i]; // 마지막 그룹의 P값을 tmp에 저장해준다.
		}
		if (flag)
			answer = (answer < sum + tmp) ? sum + tmp : answer;

		return;
	}

	for (int i = init; i < N; i++) {
		if (!check[i]) {
			check[i] = 1;
			tmp *= sakura[i]; // 벚꽃을 모두 같은 그룹으로 생각하고 계속 곱해준다.
			bruteForce(i + 1, cnt + 1, sum + tmp); // 그룹을 나눈다. 나눌 때마다 +연산을 해준다.
			check[i] = 0;
		}
	}
}

int main() {
	cin >> N;
	for (int i = 0; i < N; i++) {
		cin >> sakura[i];
	}

	bruteForce(0, 0, 0);

	cout << answer << '\n';

	return 0;
}

File: BOJ/python/17127.py
import sys

input = sys.stdin.readline


def bruteforce(init: int, cnt: int, sum: int):
    """
    4개의 그룹으로 나눌 수 있는 모든 경우의 수를 탐색하고
    그 중 가장 큰 값을 answer에 저장 한다.

    Args:
        init (int): 그룹의 시작 지점.
        cnt (int): 그룹을 나눈 횟수 (e.g. 3번 나누면 4개의 그룹이 생성)
        sum (int): 4개 그룹의 곱을 더해준 값(문제에서는 이 값을 P라고 한다.)을 구한다.
    """

    global N, cb, check, answer

    tmp = 1

    # 그룹을 3번 나누었다는 것은 4개의 그룹이 만들어 졌다는 의미이다.
    if cnt == 3:
        is_correct = False
        for i in range(init, N):
            is_correct = True  # 그룹을 정확히 4개의 그룹으로 나누었는지 체크한다.
            tmp *= cb[i]  # 마지막 그룹의 P값을 tmp에 저장해준다.
        if is_correct:
            answer = max(answer, sum + tmp)
        return

    for i in range(init, N):
        if not check[i]:
            check[i] = 1
            tmp *= cb[i]  # 우선 모두 같은 그룹으로 가정하고 계속 곱해준다.
            bruteforce(i + 1, cnt + 1, sum + tmp)  # 현재 위치에서 그룹을 나누어 본다. 나눌 때마다 '+'연산
            check[i] = 0


global N, cb, check, answer
answer = 0
N = int(input().rstrip("\n"))
cb = list(map(int, input().rstrip("\n").split(" ")))
check = [0 for _ in range(N)]

bruteforce(0, 0, 0)
print(answer)
